### PR TITLE
loadgen: generate loadgen on unique source accounts

### PR DESCRIPTION
**Pull request by marta-lokhova**
Partially resolves https://github.com/stellar/stellar-core/issues/3812

Subset of changes from https://github.com/stellar/stellar-core/pull/3787 so we can run this through CI and merge independently

### CODE REVIEW

File: src/simulation/LoadGenerator.cpp [hunk 1]
```diff
@@ -261,6 +349,15 @@ LoadGenerator::generateLoad(GeneratedLoadConfig cfg)
     }
 
     auto txPerStep = getTxPerStep(cfg.txRate, cfg.spikeInterval, cfg.spikeSize);
+    if (cfg.mode == LoadGenMode::CREATE)
+    {
+        // Limit creation to the number of accounts we have. This is only the
+        // case at the very beginning, when only root account is available for
+        // account creation
+        size_t expectedSize =
+            mInitiaAccountsCreated ? mAccountsAvailable.size() : 1;
```
Comment (dmkozh) [inline src/simulation/LoadGenerator.cpp:358]: typo: mInitialAccountsCreated

File: src/simulation/LoadGenerator.cpp [hunk 2]
```diff
@@ -515,20 +639,24 @@ LoadGenerator::updateMinBalance()
...
         auto account = TestAccount{mApp, txtest::getAccount(name.c_str()), sn};
-        ops.push_back(
-            txtest::createAccount(account.getPublicKey(), mMinBalance * 100));
+        ops.push_back(txtest::createAccount(account.getPublicKey(), balance));
 
         // Cache newly created account
-        mAccounts.insert(std::pair<uint64_t, TestAccountPtr>(
-            i, make_shared<TestAccount>(account)));
+        auto acc = make_shared<TestAccount>(account);
+        mAccounts.insert(std::pair(i, acc));
```
Comment (dmkozh) [inline src/simulation/LoadGenerator.cpp:655]: nit: `.emplace(i, acc)` (same below)

**Review by SirTyson — APPROVED** [review]
LGTM

File: src/herder/test/UpgradesTests.cpp [hunk 3]
```diff
@@ -3025,9 +3026,9 @@ TEST_CASE("upgrade to generalized tx set in network", "[upgrades][overlay]")
     }
 
     auto& loadGen = nodes[0]->getLoadGenerator();
-    // Generate 8 ledgers worth of txs (40 / 5).
+    // Generate 8 ledgers worth of txs (600 / (15*5)).
     loadGen.generateLoad(GeneratedLoadConfig::createAccountsLoad(
-        /* nAccounts */ 40, /* txRate */ 1, /* batchSize */ 1));
+        /* nAccounts */ 600, /* txRate */ 1, /* batchSize */ 15));
```
Comment (mbsdf) [inline src/herder/test/UpgradesTests.cpp:3031]: is it 600 because you want to generate 600 transactions?

File: src/simulation/LoadGenerator.cpp [hunk 4]
```diff
@@ -168,34 +170,84 @@ void
...
     if (mFailed)
     {
-        CLOG_ERROR(LoadGen, "Load generation failed, ensure correct "
-                            "number parameters are set and accounts are "
-                            "created, or retry with smaller tx rate.");
+        errorMsg = "Load generation failed, ensure correct "
+                   "number parameters are set and accounts are "
+                   "created, or retry with smaller tx rate.";
+    }
+    // During account creation, we can create up to
```
Comment (mbsdf) [inline src/simulation/LoadGenerator.cpp:199]: Can we add a comment above explaining nAccount, txRate, and batchSize relationship briefly?

File: src/simulation/LoadGenerator.cpp [hunk 5]
```diff
@@ -168,34 +170,84 @@ void
...
+    mAccountsInUse.clear();
+    mAccountsAvailable.clear();
+    mCreationSourceAccounts.clear();
+    mLoadTimer.reset();
     mRoot.reset();
     mStartTime.reset();
     mTotalSubmitted = 0;
     mWaitTillCompleteForLedgers = 0;
     mFailed = false;
+    mStarted = false;
```
Comment (mbsdf) [inline src/simulation/LoadGenerator.cpp:182]: +1  mStarted defined here

File: src/simulation/LoadGenerator.cpp [hunk 6]
```diff
@@ -969,7 +1062,7 @@ LoadGenerator::execute(TransactionFramePtr& txf, LoadGenMode mode,
     switch (mode)
     {
     case LoadGenMode::CREATE:
-        txm.mAccountCreated.Mark(batchSize);
+        txm.mAccountCreated.Mark(txf->getNumOperations());
```
Comment (mbsdf) [inline src/simulation/LoadGenerator.cpp:1065]: mode Pay is numOperations, is there something better than batchSize for Create? I am not following why need this change?

File: src/simulation/LoadGenerator.cpp [hunk 7]
```diff
@@ -168,34 +170,84 @@ void
 LoadGenerator::reset()
 {
     mAccounts.clear();
+    mAccountsInUse.clear();
```
Comment (mbsdf) [inline src/simulation/LoadGenerator.cpp:173]: can you add a few lines of comments explaining these three

File: src/simulation/LoadGenerator.h [hunk 8]
```diff
@@ -144,10 +145,24 @@ class LoadGenerator
     // Accounts cache
     std::map<uint64_t, TestAccountPtr> mAccounts;
 
+    // Track account IDs that are currently being referenced by the transaction
```
Comment (mbsdf) [inline src/simulation/LoadGenerator.h:148]: +1

File: src/simulation/LoadGenerator.cpp (hunk 7 above)
Comment (mbsdf) [reply src/simulation/LoadGenerator.cpp:173]: ok, already there in loadgenerator.h

**Review by mbsdf — COMMENTED** [review]
I have couple questions and comments. All looks fine.

File: src/simulation/LoadGenerator.cpp [hunk 9]
```diff
@@ -168,34 +170,84 @@ void
...
+                   "number parameters are set and accounts are "
+                   "created, or retry with smaller tx rate.";
+    }
+    // During account creation, we can create up to
+    // MAX_OPS_PER_TX*MAX_OPS_PER_TX new accounts per ledger. Ensure desired
+    // tx rate is sufficient, and allows some buffer in case ledgers take
+    // longer than usual.
+    else if (cfg.mode == LoadGenMode::CREATE && cfg.nAccounts > cfg.batchSize &&
+             (cfg.txRate * Herder::EXP_LEDGER_TIMESPAN_SECONDS.count() *
+              MIN_UNIQUE_ACCOUNT_MULTIPLIER) > cfg.batchSize)
```
Comment (mbsdf) [inline src/simulation/LoadGenerator.cpp:205]: batchSize?

File: src/simulation/LoadGenerator.cpp (hunk 2 above)
Comment (marta-lokhova) [reply src/simulation/LoadGenerator.cpp:655]: sure, this was copy pasta from the code in master

File: src/overlay/test/OverlayTests.cpp [hunk 10]
```diff
@@ -2619,9 +2622,9 @@ TEST_CASE("overlay pull mode loadgen", "[overlay][pullmode][acceptance]")
     // Set a really high tx rate so we create the txns right away.
     auto const numAccounts = 5;
     loadGen.generateLoad(GeneratedLoadConfig::createAccountsLoad(
-        /* nAccounts */ numAccounts,
-        /* txRate */ 1000,
-        /* batchSize */ 1));
+        /* nAccounts */ numAccounts * 15,
```
[the remark this answers is not in the document]
Comment (mbsdf) [reply src/overlay/test/OverlayTests.cpp]: +1

**Review by mbsdf — APPROVED** [review]

**Review by bitcoinfinancier — APPROVED** [review]

**Comment by dmkozh** [comment]
r+ 3732c8f

File: src/main/test/ApplicationUtilsTests.cpp [hunk 11]
```diff
@@ -214,7 +214,7 @@ class SimulationHelper
 
         // Generate a bit of load, and crank for some time
         loadGen.generateLoad(GeneratedLoadConfig::txLoad(
-            LoadGenMode::PAY, /* nAccounts */ 10, /* nTxs */ 10,
+            LoadGenMode::PAY, /* nAccounts */ 150, /* nTxs */ 10,
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/main/test/ApplicationUtilsTests.cpp]: Discussed offline for this tests and the ones below. We need more accounts in some tests to support the desired tx rate when we actually generate load. This also means that in some cases we need a higher TESTING_UPGRADE_MAX_TX_SET_SIZE in order to close ledgers with the desired load.

**Pull request merged**